src/libertem/io/dataset/hdf5.py

```diff
@@ -291,6 +349,133 @@ def _get_tiles_normal(self, tiling_scheme, tileshape_nd, dest_dtype="float32"):
+                    else:
+                        s = tuple(a + 1 for a in np.unravel_index(
+                            self.meta.image_count - o_pos - 1, ds_shape_nav)
+                        ) + tuple(s_sig)
+                        tile_slice = Slice(
+                            shape=(Shape(s, sig_dims=sig_dims)), origin=tile_slice.origin
+                        )
+                # actual reshaping of the tile_slice
+                tile_slice = tile_slice.reshape(self.meta.shape, self.native_shape)
+                # check if tile_slice needs to split, then yield the slice/slices
+                yield from self._split_and_get_tile_slice(
+                    tile_slice, axis_size, tileshape_nd, scheme_idx
+                )
+
+    def _split_and_get_tile_slice(self, tile_slice, axis_size, tileshape_nd, scheme_idx):
+        # recursive function to split a tile until it fits the axis, then yield the slice/slices
+        sig_dims = self.native_shape.sig.dims
+        o_nav = tile_slice.origin[:-sig_dims]
+        s_nav = tile_slice.shape[:-sig_dims]
+        o_sig = tile_slice.origin[-sig_dims:]
+        s_sig = tile_slice.shape[-sig_dims:]
+        ds_shape_nav = self.native_shape[:-sig_dims]
+        s_nav = tuple(a - 1 for a in s_nav)
+        o_pos = np.ravel_multi_index(o_nav, ds_shape_nav) % axis_size
+        o_pos_actual = np.ravel_multi_index(o_nav, ds_shape_nav)
+        s_pos_actual = np.ravel_multi_index(s_nav, ds_shape_nav)
+        frame_pos = o_pos + s_pos_actual
+        axis_size_pos = axis_size - 1
+        # if frame lies outside the axis size, we've to split
+        if frame_pos > axis_size_pos:
+            # first split in tile_slice
+            o_1 = tile_slice.origin
+            s_1 = tuple(a + 1 for a in np.unravel_index(
+                axis_size_pos - o_pos, ds_shape_nav)
+            ) + tuple(s_sig)
+            # second split in tile_slice
+            o_2 = np.unravel_index(
+                o_pos_actual + axis_size_pos - o_pos + 1, ds_shape_nav
+            ) + tuple(o_sig)
+            s_2 = tuple(a + 1 for a in np.unravel_index(
+                frame_pos - axis_size_pos - 1, ds_shape_nav)
+            ) + tuple(s_sig)
+            tile_slice_1 = Slice(shape=(Shape(s_1, sig_dims=sig_dims)), origin=o_1)
+            tile_slice_2 = Slice(shape=(Shape(s_2, sig_dims=sig_dims)), origin=o_2)
+            # check if tile_slice_1 needs to be split, else yield it
+            yield from self._split_and_get_tile_slice(
+                tile_slice_1, axis_size, tileshape_nd, scheme_idx
+            )
+            # check if tile_slice_2 needs to be split, else yield it
+            yield from self._split_and_get_tile_slice(
+                tile_slice_2, axis_size, tileshape_nd, scheme_idx
+            )
+        # yield the tile_slice since it fits inside the axis size
+        else:
+            with self.reader.get_h5ds() as dataset:
```

Comment:
As commented above, `dataset` should be opened once and passed down to this place.